# è½¬å½•ç»„å·®å¼‚åˆ†æé‡‘æ ‡å‡†-Limma-voomå®æˆ˜

> åˆ˜å°æ³½å†™äº19.3.11
>
> Limmaä½œä¸ºå·®å¼‚åˆ†æçš„â€œé‡‘æ ‡å‡†â€æœ€åˆæ˜¯åº”ç”¨åœ¨èŠ¯ç‰‡æ•°æ®åˆ†æä¸­ï¼Œvoomçš„åŠŸèƒ½æ˜¯ä¸ºäº†RNA-Seqçš„åˆ†æäº§ç”Ÿçš„ã€‚è¯¦ç»†æ¢ç´¢ä¸€ä¸‹limmaçš„åŠŸèƒ½å§
>
> æœ¬æ¬¡çš„æµ‹è¯•æ•°æ®å¯ä»¥åœ¨**å…¬ä¼—å·å›å¤voom**è·å¾—

Limma-voomå¼ºå¤§åœ¨äºä¸‰ä¸ªæ–¹é¢ï¼š

- False discovery rateæ¯”è¾ƒä½ï¼ˆå‡†ç¡®æ€§ï¼‰ï¼Œå¼‚å¸¸å€¼å½±å“å°
- å‡é˜³æ€§æ§åˆ¶ä¸é”™
- è¿ç®—å¾ˆå¿«

### é…ç½®ä¿¡æ¯

```R
> library(edgeR)
Loading required package: limma
> counts <- read.delim("all_counts.txt", row.names = 1)
> head(counts[1:3,1:3])
          C61 C62 C63
AT1G01010 341 371 275
AT1G01020 164  94 176
AT1G03987   0   0   0
> dim(counts)
[1] 32833    24
# æ„å»ºDGEListå¯¹è±¡ï¼Œå°†countså’Œsampleä¿¡æ¯åŒ…å«è¿›å»
> d0 <- DGEList(counts)
```

### é¢„å¤„ç†

```R
> # è®¡ç®—æ ‡å‡†åŒ–å› å­
> d0 <- calcNormFactors(d0)
> d0
```

![](https://upload-images.jianshu.io/upload_images/9376801-761a1695ee8c2287.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

æ³¨æ„ï¼Œè¿™é‡Œçš„`calcNormFactors`å¹¶ä¸æ˜¯è¿›è¡Œäº†æ ‡å‡†åŒ–ï¼Œä»…ä»…æ˜¯è®¡ç®—äº†ä¸€ä¸ªå‚æ•°ï¼Œç”¨äºä¸‹æ¸¸æ ‡å‡†åŒ–

```R
# è¿‡æ»¤ä½è¡¨è¾¾åŸºå› [é˜ˆå€¼æ ¹æ®è‡ªå·±éœ€è¦è®¾å®š]
cutoff <- 1
cut <- which(apply(cpm(d0), 1, max) < cutoff)
d <- d0[-cut,] 
dim(d) 
[1] 21081    24
# å‰©ä¸‹ 21081ä¸ªåŸºå› 
```

ç„¶åæ ¹æ®åˆ—åæå–æ ·æœ¬ä¿¡æ¯(sample name)

```R
> spname <- colnames(counts) 
> spname
 [1] "C61"  "C62"  "C63"  "C64"  "C91"  "C92"  "C93" 
 [8] "C94"  "I561" "I562" "I563" "I564" "I591" "I592"
[15] "I593" "I594" "I861" "I862" "I863" "I864" "I891"
[22] "I892" "I893" "I894"
```

çœ‹åˆ°æ ·æœ¬æ˜¯æŒ‰ç…§ä¸¤ä¸ªå› ç´ ï¼ˆå“ç³»C/I5/I8ã€æ—¶é—´6/9ï¼‰åˆ†ç±»çš„ï¼Œå¹¶ä¸”å››ä¸ªç”Ÿç‰©å­¦é‡å¤å†™åœ¨äº†æœ€å`C/I5/I8 | 6/9 | 1/2/3/4` 

```R
> # åˆ†ç¦»å‡ºåˆ†ç»„ä¿¡æ¯
> strain <- substr(spname, 1, nchar(spname) - 2)
> time <- substr(spname, nchar(spname) - 1, nchar(spname) - 1)
> strain
 [1] "C"  "C"  "C"  "C"  "C"  "C"  "C"  "C"  "I5" "I5"
[11] "I5" "I5" "I5" "I5" "I5" "I5" "I8" "I8" "I8" "I8"
[21] "I8" "I8" "I8" "I8"
> time
 [1] "6" "6" "6" "6" "9" "9" "9" "9" "6" "6" "6" "6"
[13] "9" "9" "9" "9" "6" "6" "6" "6" "9" "9" "9" "9"
```

å†å°†è¿™ä¸¤éƒ¨åˆ†æ•´åˆè¿›groupåˆ†ç»„ä¿¡æ¯ä¸­

```R
> # å†å°†è¿™ä¸¤éƒ¨åˆ†æ•´åˆè¿›groupåˆ†ç»„ä¿¡æ¯ä¸­[ç”Ÿæˆå› å­å‹å‘é‡]
> group <- interaction(strain, time)
> group
 [1] C.6  C.6  C.6  C.6  C.9  C.9  C.9  C.9  I5.6 I5.6
[11] I5.6 I5.6 I5.9 I5.9 I5.9 I5.9 I8.6 I8.6 I8.6 I8.6
[21] I8.9 I8.9 I8.9 I8.9
Levels: C.6 I5.6 I8.6 C.9 I5.9 I8.9
```

å½“ç„¶ï¼Œä¹Ÿå¯ä»¥è‡ªå·±æ‰‹åŠ¨è¾“å…¥æˆ–ä»å…¶ä»–æ–‡ä»¶å¯¼å…¥ï¼Œä½†**å¿…é¡»æ³¨æ„ä¸€ç‚¹**ï¼šè¿™ä¸ªgroup metadataå¿…é¡»å’Œcountsçš„åˆ—æ˜é¡ºåºå¯¹åº”

å¤šä¸ªå®éªŒå› å­åŒæ—¶å­˜åœ¨æ—¶ï¼Œè¦è¿›è¡ŒMDSï¼ˆmultidimensional scalingï¼‰åˆ†æï¼Œå³â€œå¤šç»´å°ºåº¦å˜æ¢â€ã€‚æ­£å¼å·®å¼‚åˆ†æå‰å¸®åŠ©æˆ‘ä»¬åˆ¤æ–­æ½œåœ¨çš„å·®å¼‚æ ·æœ¬ï¼Œç»“æœä¼šå°†æ‰€æœ‰æ ·æœ¬åˆ’åˆ†æˆå‡ ä¸ªç»´åº¦ï¼Œç¬¬ä¸€ç»´åº¦çš„æ ·æœ¬ä»£è¡¨äº†æœ€å¤§çš„å·®å¼‚

```R
> # Multidimensional scaling (MDS) plot
> suppressMessages(library(RColorBrewer))
> col.group <- group
> levels(col.group) <- brewer.pal(nlevels(col.group), "Set1") 
> col.group <- as.character(col.group)
> plotMDS(d, labels=group, col=col.group) 
> title(main="A. Sample groups")
```

![](https://upload-images.jianshu.io/upload_images/9376801-3cd60a9c65f7bbfb.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

### Voomè½¬æ¢åŠæ–¹å·®æƒé‡è®¡ç®—

```R
> mm <- model.matrix(~0 + group)
> y <- voom(d, mm, plot = T)
```

![Good](https://upload-images.jianshu.io/upload_images/9376801-bd482cd4342d26ea.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

> **voomåˆ°åº•åšäº†ä»€ä¹ˆè½¬æ¢ï¼Ÿ**

é¦–å…ˆåŸå§‹countsè½¬æ¢æˆlog2çš„CPMï¼ˆcounts per million reads ï¼‰ï¼Œè¿™é‡Œçš„per million readsæ˜¯æ ¹æ®ä¹‹å‰`calcNormFactors`è®¡ç®—çš„norm.factorsè¿›è¡Œè§„å®šçš„ï¼›

ç„¶åæ ¹æ®æ¯ä¸ªåŸºå› çš„log2CPMåˆ¶ä½œäº†çº¿æ€§æ¨¡å‹ï¼Œå¹¶è®¡ç®—äº†[æ®‹å·®](https://stattrek.com/statistics/dictionary.aspx?definition=residual) ï¼›

ç„¶ååˆ©ç”¨äº†å¹³å‡è¡¨è¾¾é‡ï¼ˆçº¢çº¿ï¼‰æ‹Ÿåˆäº†sqrt(residual standard deviation)ï¼›

æœ€åå¾—åˆ°çš„å¹³æ»‘æ›²çº¿å¯ä»¥ç”¨æ¥å¾—åˆ°æ¯ä¸ªåŸºå› å’Œæ ·æœ¬çš„æƒé‡

https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29

**ä¸Šå›¾æ•ˆæœè¾ƒå¥½ï¼Œå¦‚æœåƒä¸‹é¢ğŸ‘‡è¿™æ ·ï¼šå°±è¡¨ç¤ºæ•°æ®éœ€è¦å†è¿›è¡Œè¿‡æ»¤**

```R
tmp <- voom(d0, mm, plot = T)
```

![Bad](https://upload-images.jianshu.io/upload_images/9376801-b61fbb8641b2f214.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

> æœ‰æ—¶æˆ‘ä»¬æ²¡æœ‰å¿…è¦å¼„æ˜ç™½èƒŒåå¤æ‚çš„åŸç†ï¼Œåªéœ€è¦çŸ¥é“å¦‚ä½•è§£é‡Šç»“æœï¼š
>
> https://stats.stackexchange.com/questions/160255/voom-mean-variance-trend-plot-how-to-interpret-the-plot
>
> limma-voom method assumes that rows with zero or very low counts have been removed
>
> å¦‚æœæ¨ªåæ ‡æ¥è¿‘0çš„ä½ç½®å‡ºç°è¿…é€Ÿä¸Šå‡ï¼Œè¯´æ˜low countsæ•°æ¯”è¾ƒå¤š
>
> Whether your data are "good" or not cannot be determined from this plot

### limmaçš„çº¿æ€§æ‹Ÿåˆæ¨¡å‹æ„å»º

```R
> fit <- lmFit(y, mm)
> head(coef(fit),3)
          groupC.6 groupI5.6 groupI8.6 groupC.9
AT1G01010 4.685920 5.2477564  4.938939 4.922501
AT1G01020 3.420726 3.4147535  3.130644 3.571855
AT1G01030 1.111114 0.7316936  1.435521 1.157532
          groupI5.9 groupI8.9
AT1G01010  5.382619  5.246093
AT1G01020  3.610579  3.655254
AT1G01030  0.388736  1.222892
```

ç»„é—´æ¯”è¾ƒï¼š

ä¾‹å¦‚è¿›è¡ŒI5å“ç³»çš„6å’Œ9å°æ—¶æ¯”è¾ƒ

```R
> contr <- makeContrasts(groupI5.9 - groupI5.6, levels = colnames(coef(fit)))
> contr
           Contrasts
Levels      groupI5.9 - groupI5.6
  groupC.6                      0
  groupI5.6                    -1
  groupI8.6                     0
  groupC.9                      0
  groupI5.9                     1
  groupI8.9                     0
```

ä¼°ç®—ç»„é—´æ¯ä¸ªåŸºå› çš„æ¯”è¾ƒï¼š

```R
> tmp <- contrasts.fit(fit, contr)
```

å†åˆ©ç”¨Empirical Bayes ï¼ˆshrinks standard errors that are much larger or smaller than those from other genes towards the average standard erroï¼‰

https://www.degruyter.com/doi/10.2202/1544-6115.1027

```R
> tmp <- eBayes(tmp)
```

### ç»“æœä¸­å·®å¼‚åŸºå› æœ‰å“ªäº›å‘¢ï¼Ÿ

```R
> top.table <- topTable(tmp, sort.by = "P", n = Inf)
> DEG <- na.omit(top.table)
> head(DEG, 5)
              logFC  AveExpr         t      P.Value
AT5G37260  3.163518 6.939588  23.94081 1.437434e-16
AT3G02990  1.646438 3.190750  13.15656 1.610004e-11
AT2G29500 -5.288998 5.471250 -11.94053 9.584101e-11
AT3G24520  1.906690 5.780286  11.80461 1.179985e-10
AT5G65630  1.070550 7.455294  10.86740 5.208111e-10
             adj.P.Val        B
AT5G37260 3.030255e-12 26.41860
AT3G02990 1.697025e-07 15.50989
AT2G29500 6.218818e-07 14.45441
AT3G24520 6.218818e-07 14.52721
AT5G65630 2.195844e-06 13.12701
```

- logFC: log2 fold change of **I5.9/I5.6**
- AveExpr: Average expression across all samples, in log2 CPM
- t: logFC divided by its standard error
- P.Value: Raw p-value (based on t) from test that logFC differs from 0
- adj.P.Val: Benjamini-Hochberg false discovery rate adjusted p-value
- B: log-odds that gene is DE (arguably less useful than the other columns)

ä»å‰å‡ ä¸ªå·®å¼‚æœ€æ˜¾è‘—çš„åŸºå› ä¸­å¯ä»¥çœ‹åˆ°ï¼ŒAT5G37260åŸºå› åœ¨time9çš„è¡¨è¾¾é‡æœ€é«˜ï¼ˆçº¦time6çš„8å€ï¼‰ï¼ŒAT2G29500è¡¨è¾¾é‡æœ€ä½ï¼Œæ¯”time6çš„è¿˜å°‘ï¼ˆçº¦1/32ï¼‰

é‚£ä¹ˆæ€»å…±æœ‰å¤šå°‘å·®å¼‚åŸºå› å‘¢ï¼Ÿ

å¦‚æœä»¥logFC=2ï¼ŒPvalue=0.05ä¸ºé˜ˆå€¼è¿›è¡Œè¿‡æ»¤

```R
> length(which(DEG$adj.P.Val < 0.05 & abs(DEG$logFC)>2 ))
[1] 172
```

>  å¦‚æœè¦æ¯”è¾ƒå…¶ä»–çš„ç»„ï¼Œä¾‹å¦‚ï¼štime6çš„å“ç³»Cå’Œå“ç³»I5

åªéœ€è¦å°†`makeContrasts`ä¿®æ”¹

```R
contr <- makeContrasts(groupI5.6 - groupC.6, levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
top.table <- topTable(tmp, sort.by = "P", n = Inf)
DEG <- na.omit(top.table)
head(DEG, 5)
length(which(DEG$adj.P.Val < 0.05 & abs(DEG$logFC)>2 ))
# ç»“æœåªæœ‰8ä¸ª
```

---

ä¸Šé¢åˆ©ç”¨äº†å•å› å­groupæ„å»ºäº†model matrixï¼Œå¦‚æœå­˜åœ¨å¤šä¸ªå½±å“å› å­ï¼Œå¯ä»¥åˆ©ç”¨æ–°çš„å› å­ï¼ˆå°±çœå»äº†ä¹‹å‰å› å­ç»„åˆæˆgroupçš„æ­¥éª¤ï¼‰æ„å»ºæ–°çš„çŸ©é˜µæ¨¡å‹

```R
# æ„å»ºæ–°çš„model matrix
> mm <- model.matrix(~strain*time)
> colnames(mm)
[1] "(Intercept)"    "strainI5"       "strainI8"      
[4] "time9"          "strainI5:time9" "strainI8:time9"
> y <- voom(d, mm, plot = F)
> fit <- lmFit(y, mm)
> head(coef(fit),3)
          (Intercept)     strainI5   strainI8
AT1G01010    4.685920  0.561836365  0.2530188
AT1G01020    3.420726 -0.005972208 -0.2900818
AT1G01030    1.111114 -0.379420605  0.3244063
              time9 strainI5:time9 strainI8:time9
AT1G01010 0.2365808    -0.10171813     0.07057368
AT1G01020 0.1511295     0.04469623     0.37348052
AT1G01030 0.0464182    -0.38937581    -0.25904674
```

- ç®—æ³•è‡ªå®šä¹‰äº†æ ‡å‡†å“ç³»ä¸ºCï¼Œæ ‡å‡†æ—¶é—´ä¸º6ï¼ˆå¯èƒ½æ˜¯æŒ‰ç…§å­—æ¯æˆ–æ•°å­—é¡ºåºï¼‰
- `strainI5`è¡¨ç¤ºå“ç³»I5å’Œæ ‡å‡†å“ç³»ï¼ˆå“ç³»Cï¼‰åœ¨æ ‡å‡†æ—¶é—´ç‚¹ï¼ˆtime6ï¼‰çš„å·®å¼‚
- `time9`è¡¨ç¤ºæ ‡å‡†å“ç³»ï¼ˆå“ç³»Cï¼‰åœ¨time9å’Œtime6çš„å·®å¼‚
- `strainI5:time9`è¡¨ç¤ºå“ç³»I5å’Œå“ç³»Cåœ¨time9å’Œtime6çš„å·®å¼‚ï¼ˆå­˜åœ¨äº¤å‰å½±å“ï¼‰

å¦‚æœæˆ‘ä»¬æƒ³**æ¯”è¾ƒå“ç³»I5å’Œå“ç³»Cåœ¨time6çš„å·®å¼‚**ï¼Œå°±å¯ä»¥ï¼š

```R
> tmp <- contrasts.fit(fit, coef = 2)
> tmp <- eBayes(tmp)
> top.table <- topTable(tmp, sort.by = "P", n = Inf)
> DEG <- na.omit(top.table)
> head(DEG, 5)
               logFC    AveExpr          t
AT4G12520 -10.254556  0.3581132 -11.402477
AT3G30720   5.817438  3.3950689  10.528934
AT5G26270   2.421030  4.3788335   9.654257
AT3G33528  -4.780814 -1.8612945  -7.454943
AT1G64795  -4.872595 -1.3119360  -7.079643
               P.Value    adj.P.Val          B
AT4G12520 2.206726e-10 4.651998e-06  3.6958152
AT3G30720 9.108689e-10 9.601014e-06  7.9963406
AT5G26270 4.101051e-09 2.881809e-05 10.8356224
AT3G33528 2.741289e-07 1.444728e-03  0.5677732
AT1G64795 5.985471e-07 2.523594e-03  1.8151705
> length(which(DEG$adj.P.Val < 0.05 & abs(DEG$logFC)>2 ))
[1] 8
```

å¯ä»¥çœ‹åˆ°ï¼Œå’Œä¹‹å‰ç”¨å•å› å­groupå¾—åˆ°çš„ç»“æœä¸€æ ·

ä½†æ˜¯ï¼Œè¿™ç§æ–¹æ³•åœ¨åŒæ—¶**åˆ†æäº¤å‰å½±å“æ—¶å°±ä½“ç°å‡ºæ¥å¼ºå¤§äº†ï¼š**

æ¯”å¦‚æˆ‘ä»¬**æƒ³çœ‹time9ä¸å“ç³»I5çš„å·®å¼‚ç»“æœ**

![](https://upload-images.jianshu.io/upload_images/9376801-5f860d36c34e0361.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

```R
> # cultivarI5:time9
> tmp <- contrasts.fit(fit, coef = 5)
> tmp <- eBayes(tmp)
> top.table <- topTable(tmp, sort.by = "P", n = Inf)
> DEG <- na.omit(top.table)
> #head(DEG, 5)
> length(which(DEG$adj.P.Val < 0.05 & abs(DEG$logFC)>2 ))
[1] 111
```

### æ›´å¤æ‚çš„æ¨¡å‹

æœ‰æ—¶RNA-Seqéœ€è¦è€ƒè™‘**æ‰¹æ¬¡æ•ˆåº”ï¼ˆBatch effectï¼‰çš„å½±å“**

```R
> batch <- factor(rep(rep(1:2, each = 2), 6))
> batch
 [1] 1 1 2 2 1 1 2 2 1 1 2 2 1 1 2 2 1 1 2 2 1 1 2 2
Levels: 1 2
```

æ„å»ºæ¨¡å‹æ—¶ï¼Œéœ€è¦å°†batchåŠ åœ¨æœ€åï¼Œå…¶ä»–ä¸å˜

```R
> mm <- model.matrix(~0 + group + batch)
> y <- voom(d, mm, plot = F)
> fit <- lmFit(y, mm)
> contr <- makeContrasts(groupI5.6 - groupC.6, levels = colnames(coef(fit)))
> tmp <- contrasts.fit(fit, contr)
> tmp <- eBayes(tmp)
> top.table <- topTable(tmp, sort.by = "P", n = Inf)
> DEG <- na.omit(top.table)
> #head(DEG, 5)
> length(which(DEG$adj.P.Val < 0.05 & abs(DEG$logFC)>2 ))
[1] 9
```



æˆ–è€…éœ€è¦è€ƒè™‘å…¶ä»–å› ç´ çš„å½±å“ï¼Œæ¯”å¦‚è¿™é‡Œæœ‰**ä¸€ä¸ªè¿ç»­å‹å˜é‡rateï¼Œå®ƒå¯èƒ½æ˜¯pHã€å…‰ç…§ç­‰ç­‰å¯¹ç ”ç©¶ææ–™çš„å½±å“å€¼**

```R
> # Generate example rate data[è¡Œæ•°è¦ä¸countçŸ©é˜µçš„åˆ—æ•°ç›¸ç­‰]
> set.seed(10)
> rate <- rnorm(n = 24, mean = 5, sd = 1.7)
> rate
 [1] 5.031868 4.686771 2.668738 3.981415 5.500727
 [6] 5.662650 2.946271 4.381751 2.234656 4.563987
[11] 6.873025 6.284829 4.595003 6.678656 6.260363
[16] 5.151890 3.376595 4.668244 6.573386 5.821063
> # æŒ‡å®šçŸ©é˜µæ¨¡å‹
> mm <- model.matrix(~rate)
> head(mm)
  (Intercept)     rate
1           1 5.031868
2           1 4.686771
3           1 2.668738
4           1 3.981415
5           1 5.500727
6           1 5.662650
> y <- voom(d, mm, plot = F)
> fit <- lmFit(y, mm)
> tmp <- contrasts.fit(fit, coef = 2) # test "rate" coefficient
> tmp <- eBayes(tmp)
> top.table <- topTable(tmp, sort.by = "P", n = Inf)
> DEG <- na.omit(top.table)
> #head(DEG, 5)
> length(which(DEG$adj.P.Val < 0.05 & abs(DEG$logFC)>2 ))
[1] 0
```

å¯è§rateå€¼å¹¶ä¸èƒ½æˆä¸ºäº§ç”Ÿå·®å¼‚åŸºå› çš„åŸå› ï¼Œä½†æ˜¯**rateä¸åŸºå› çš„ç›¸å…³æ€§è¿˜æ˜¯å¯ä»¥æ¢ç´¢ä¸€ä¸‹çš„**

```R
> AT1G01060 <- y$E["AT1G01060",]
> plot(AT1G01060 ~ rate, ylim = c(6, 12))
> intercept <- coef(fit)["AT1G01060", "(Intercept)"]
> slope <- coef(fit)["AT1G01060", "rate"]
> abline(a = intercept, b = slope)
```

![](https://upload-images.jianshu.io/upload_images/9376801-9c02b8c015a443c1.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

å›¾ä¸­çš„**æ–œç‡å°±æ˜¯logFCå€¼**ï¼Œæˆ–è€…å¯ä»¥è¯´æ¯å•ä½rateçš„å¢åŠ ï¼Œgeneè¡¨è¾¾é‡log2 CPMçš„æ”¹å˜ã€‚è¿™é‡Œæ–œç‡ä¸º-0.096è¡¨ç¤ºï¼šæ¯å•ä½rateçš„å¢åŠ ï¼Œå°±æœ‰-0.096 log2CPMçš„åŸºå› è¡¨è¾¾é‡é™ä½ï¼›æˆ–è€…æ¯å•ä½rateçš„å¢åŠ ï¼Œå°±æœ‰6.9%çš„CPMé™ä½ï¼ˆ`2^0.096 = 1.069`ï¼‰ 